### ISSUE TITLE
JsonSerializerOptions is not accepted for JsonSerializer.Deserialize

### ISSUE THREAD

**Issue by SetTrend** [dotnet/roslyn#77474]
### Version Used:

 .NET 9.0

### Steps to Reproduce:

```c#
using System;
using System.IO;
using System.Text.Json;

namespace ClassLibrary1
{
  public class Options
  {
    private static readonly JsonSerializerOptions _jsonOptions = new JsonSerializerOptions() { IgnoreReadOnlyProperties = true, WriteIndented = true };

    public static Options Load()
    {
      using FileStream fs = new FileStream("", FileMode.Open, FileAccess.Read, FileShare.Read);

      return JsonSerializer.Deserialize(fs, _jsonOptions) ?? throw new InvalidOperationException("Cannot deserialize application configuration file. Invalid format.");
    }

  }
}
```

### Diagnostic Id:

[`CS1503`](https://learn.microsoft.com/en-us/dotnet/csharp/misc/cs1503)`: Argument 2: cannot convert from 'System.Text.Json.JsonSerializerOptions' to 'System.Text.Json.Serialization.Metadata.JsonTypeInfo'`

### Expected Behavior:

This should compile flawlessly.

### Actual Behavior:

The above error is reported.

![Cannot use JsonSerializerOptions](https://github.com/user-attachments/assets/689abb34-7773-4331-b88d-6fc91a6101a4)

**Comment by SetTrend** [comment]
I noticed that the `JsonSerializer.Deserialize()` call is accepted when I add a type argument, like

```c#
  return JsonSerializer.Deserialize<Options>(fs, _jsonOptions);
```

Shouldn't the compiler be able to infer the type from the outer function's return type? And if not, wouldn't it be a good idea to create an analyzer (and a Quick Fix action) that explains and fixes the situation?

**Comment by RikkiGibson** [comment]
> Shouldn't the compiler be able to infer the type from the outer function's return type?

This isn't the design of the language. The target type doesn't contribute to type inference of an invocation.

What's happening here is that overload resolution is failing and compiler is trying to pick one overload which seems most relevant to give an error about to get the user on the right track. It's possible we could improve the logic here so that our "best candidate" is `TValue? Deserialize<TValue>(Stream, JsonSerializerOptions)`, rather than the `Deserialize(Stream, JsonTypeInfo)` which appears to be getting picked here, which would let us give a diagnostic that a type argument is needed here.

**Comment by SetTrend** [comment]
Well, if a method's generic parameter type is solely used for the method's return value, wouldn't it make sense then to include the expected return type to inference then? There'd be no other source for inference otherwise.

**Comment by CyrusNajmabadi** [comment]
That would be a request for a language change. Such requests would need to start with a discussion over at dotnet/csharplang.  Thanks!

**Comment by SetTrend** [comment]
Ah, I see. Understood.

Thank you for pointing me to the right place, @CyrusNajmabadi 👍
